モデルの作成

MyskYko opened the pull request:
## やったこと
 - モデルの作成
 - 投稿周りの動作の作成

## 関連イシュー
#4 #5 

## 注意点

## 特にレビューして欲しいポイント
 - [ ] realmの使い方
 - [ ] publish subject周り

MyskYko committed:
created models

MyskYko committed:
create submit by realm

yukitoto commented on Miyatter/Comment.swift:
```diff
@@ -0,0 +1,21 @@
+    dynamic var id = 0
+    dynamic var date: NSDate?
+    dynamic var content = ""
+    dynamic var tweet: Tweet?
```
これいらない気がする

MyskYko replied:
すいません
Tweetを持つ必要がないということですか
これ、だと少し分かりづらいです

yukitoto commented on Miyatter/MakeTweetViewController.swift:
```diff
@@ -15,6 +16,7 @@ class MakeTweetViewController: UIViewController {
     // MARK: - Properties -
     
     fileprivate let disposeBag = DisposeBag()
+    fileprivate let viewModel = TweetViewModel()
```
viewModelのインスタンスは外部から`init`で渡す形にしよう。そっちの方がのちのちの柔軟性が高まるので。

yukitoto commented on Miyatter/TweetViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+    fileprivate let disposeBag = DisposeBag()
+    
+    
+    // MARK: - Inputs -
```
Good

yukitoto commented on Miyatter/TweetViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+        submitTweet
+            .subscribe(onNext: { (text) in
+                let realm = try! Realm()
+                let tweet = Tweet()
```
今回の場合は、新規でオブジェクトを作って、standaloneなインスタンスなので、writeトランザクションの外でプロパティをセットしてもクラッシュしないけど、今後アップデート処理をするときとかに危険なので、プロパティのセットはwriteトランザクションでするようにしよう。

yukitoto commented on Miyatter/TweetViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+                    realm.add(tweet)
+                }
+            })
+            .disposed(by: disposeBag)
```
こんな書き方あったのか。こっちのほうがいいな笑

yukitoto commented on Miyatter/TweetViewModel.swift:
```diff
@@ -0,0 +1,40 @@
+import RealmSwift
+import RxSwift
+
+class TweetViewModel {
```
overrideする可能性の無いクラスは、`final class ~`的な感じでfinalにしちゃおう。理由はprivateにするのと同じです！

yukitoto commented on Miyatter/Tweet.swift:
```diff
@@ -0,0 +1,21 @@
+    dynamic var id = 0
+    dynamic var date: NSDate?
+    dynamic var content = ""
+    let comments = LinkingObjects(fromType: Comment.self, property: "tweet")
```
`let comments = List<Comment>()`

こっちのほうが一般的だと思うんだよね。
LinkingObjectsっておそらく、とってくるのに時間がかかる処理なんだよね。`comment.tweet`と`post.comments`ってどっちのほうがアクセスする頻度が高いかって言うと圧倒的に後者の方が高くて、前者はむしろ使う機会ないぐらいだから、`let comments = List<Comment>()`こっちのほうが良さげ。

yukitoto reviewed (commented):
レビューしました〜

MyskYko committed:
modified
 - final
 - init(viewModel)
 - Instance in write_transaction
 - remove Tweet(Property) in Comment(class)
 - comments(Property) in Tweet(class) to be List<Comment>

MyskYko commented:
## やったこと
 - finalを用いた
 - initでviewModelを受け取る
 - インスタンス化をwriteトランザクション内で行う
 - CommentクラスのTweetを削除
 - TweetクラスのcommentsをListに変更
## 関連イシュー
#4 #5 

## 注意点

## 特にレビューして欲しいポイント

yukitoto commented on Miyatter/MakeTweetViewController.swift:
```diff
@@ -15,6 +16,7 @@ class MakeTweetViewController: UIViewController {
     // MARK: - Properties -
     
     fileprivate let disposeBag = DisposeBag()
+    fileprivate let viewModel: MakeTweetViewModel?
```
この形でOK！
ただオプショナルにする必要ない気がする

MyskYko committed:
modified
 - remove unnecessary optional
 - NSDate to Date and initialize in Model Class
 - add id into indexedProperties
